workload: fix TPC-H generator to generate correct values for p_name and ps_supplycost

rytaft opened the pull request:
**workload: fix TPC-H generator to generate correct values for p_name**

The TPC-H generator creates values for `p_name` in the `part` table by selecting 5 elements from a list of words called `randPartNames`. It is supposed to select 5 random unique elements from the full list, but prior to this commit, it only selected a permutation of the first 5 elements. This commit fixes the logic and adds a unit test.

Epic: None
Release note: None

**workload: fix TPC-H generator value of ps_supplycost**

Prior to this commit, the TPC-H generator was generating
a value between 0 and 10 for `ps_supplycost` in the `partsupp`
table. However, the spec calls for a random value in the range 
`[1.00 .. 1,000.00]`. This commit fixes the oversight.

Epic: None
Release note: None

rytaft committed:
workload: fix TPC-H generator to generate correct values for p_name

The TPC-H generator creates values for p_name in the part table by selecting
5 elements from a list of words called randPartNames. It is supposed to select
5 random unique elements from the full list, but prior to this commit, it only
selected a permutation of the first 5 elements. This commit fixes the logic and
adds a unit test.

Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93814)
<!-- Reviewable:end -->

rytaft committed:
workload: fix TPC-H generator value of ps_supplycost

Prior to this commit, the TPC-H generator was generating
a value between 0 and 10 for ps_supplycost in the partsupp
table. However, the spec calls for a random value in the range
[1.00 .. 1,000.00]. This commit fixes the oversight.

Epic: None
Release note: None

jordanlewis commented on pkg/workload/tpch/generate.go:
```diff
@@ -213,7 +213,7 @@ func (w *tpch) tpchPartSuppInitialRowBatch(
 		// PS_AVAILQTY random value [1 .. 9,999].
 		availQtyCol[i] = int16(randInt(rng, 1, 9999))
 		// PS_SUPPLYCOST random value [1.00 .. 1,000.00].
-		supplyCostCol[i] = float64(randFloat(rng, 1, 1000, 100))
```
This seems out of sync with the comment - should the comment be 10,000 or should the value stay 1,000? Or does this function do something odd?

rytaft replied:
The comment is correct. `randFloat` does do something a bit odd:
```
func randFloat(rng *rand.Rand, x, y, shift int) float32 {
	return float32(randInt(rng, x, y)) / float32(shift)
}
```

rytaft replied:
Just realized that I needed to change the first value too -- done.

jordanlewis replied:
👍

jordanlewis commented on pkg/workload/tpch/random.go:
```diff
@@ -129,9 +129,8 @@ func randPartName(rng *rand.Rand, namePerm []int, a *bufalloc.ByteAllocator) []b
-		namePerm[i] = namePerm[j]
-		namePerm[j] = i
+		j := rng.Intn(len(namePerm))
+		namePerm[i], namePerm[j] = namePerm[j], namePerm[i]
```
🤦
How did you notice this?

rytaft replied:
I noticed some weird values while working on a [stats PR](https://github.com/cockroachdb/cockroach/pull/68698) over a year ago (this has been on my todo list since then... haha)

jordanlewis replied:
haha, woohoo!

rytaft reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93814#-:-NJSON-vETlGJ_U4kgES:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @renatolabs and @smg260)



<!-- Sent from Reviewable.io -->

srosenberg reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93814)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @renatolabs, @rytaft, and @smg260)

___
*[`pkg/workload/tpch/random_test.go` line 46 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/93814#-NJSyMGg0HRZdTdIRU7b:-NJSyMGg0HRZdTdIRU7c:b-bo88rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c9dd1b41f721540ab61846f9d357da511e6dcb6/pkg/workload/tpch/random_test.go#L46)):*
> ```Go
> 
> 	// We can't guarantee much about the global distribution of names,
> 	// but we should make sure that we're not always using the same 5
> ```

I think we could safely make a stronger assertion. 92 choose 5 gives 4,9177,128 unique combinations. After 100 shuffles, the probability of seeing the same combination is astronomically low.
___
*[`pkg/workload/tpch/random_test.go` line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/93814#-NJSy2s31uBVeaZzbJyr:-NJSy2s31uBVeaZzbJys:brosl4q) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c9dd1b41f721540ab61846f9d357da511e6dcb6/pkg/workload/tpch/random_test.go#L49)):*
> ```Go
> 	// names. Run up to 100 times before failing.
> 	for i := 0; i < 100; i++ {
> 		if len(seen) > 5 {
> ```

`nPartNames` instead of the magic number.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93814)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @renatolabs, @rytaft, and @smg260)

___
*[`pkg/workload/tpch/random.go` line 133 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93814#-NJRk61q7mlAxgkD9gWO-r2-133:-NJcUskz0zxr50qqSfEK:b-9352d9) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c166beefdb1baa23632aff7bc0994bf3df1b4e9/pkg/workload/tpch/random.go#L133)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

haha, woohoo!
</blockquote></details>

I remember talking to Lina about this bug (which she shared with me) at some point this Summer, but we forgot to follow up.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
TFTRs! Last failure is an unrelated flake: #93460.

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93814#-:-NJf3Amr9P8huQ7xhMhm:b-tu4qat)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis, @renatolabs, @smg260, and @srosenberg)

___
*[`pkg/workload/tpch/random.go` line 133 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/93814#-NJRk61q7mlAxgkD9gWO-r2-133:-NJf37EB4odkYjfe4vQf:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/9c166beefdb1baa23632aff7bc0994bf3df1b4e9/pkg/workload/tpch/random.go#L133)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I remember talking to Lina about this bug (which she shared with me) at some point this Summer, but we forgot to follow up.
</blockquote></details>

:thumbsup: 
___
*[`pkg/workload/tpch/random_test.go` line 46 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/93814#-NJSyMGg0HRZdTdIRU7b:-NJf32nf9kNsjFH1qk57:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c9dd1b41f721540ab61846f9d357da511e6dcb6/pkg/workload/tpch/random_test.go#L46)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

I think we could safely make a stronger assertion. 92 choose 5 gives 4,9177,128 unique combinations. After 100 shuffles, the probability of seeing the same combination is astronomically low.
</blockquote></details>

Done.
___
*[`pkg/workload/tpch/random_test.go` line 49 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/93814#-NJSy2s31uBVeaZzbJyr:-NJf33BTDIu-40I9wfqi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7c9dd1b41f721540ab61846f9d357da511e6dcb6/pkg/workload/tpch/random_test.go#L49)):*
<details><summary><i>Previously, srosenberg (Stan Rosenberg) wrote…</i></summary><blockquote>

`nPartNames` instead of the magic number.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
